src/main/scala/com/microsoft/hyperspace/index/CoveringIndex.scala

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright (2021) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index
+
+import com.fasterxml.jackson.annotation.{JsonGetter, JsonRawValue}
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize
+import org.apache.spark.sql.{DataFrame, SaveMode}
+import org.apache.spark.sql.catalyst.catalog.BucketSpec
+import org.apache.spark.sql.functions.{col, input_file_name}
+import org.apache.spark.sql.hyperspace.utils.StructTypeUtils
+import org.apache.spark.sql.types.StructType
+
+import com.microsoft.hyperspace.HyperspaceException
+import com.microsoft.hyperspace.index.DataFrameWriterExtensions.Bucketizer
+import com.microsoft.hyperspace.shim.{StructTypeConverter, UntilSpark24}
+import com.microsoft.hyperspace.util.ResolverUtils
+import com.microsoft.hyperspace.util.ResolverUtils.ResolvedColumn
+
+case class CoveringIndex(
+    override val indexedColumns: Seq[String],
+    includedColumns: Seq[String],
+    // See schemaJson for more information about the annotation.
+    @JsonDeserialize(converter = classOf[StructTypeConverter.T]) schema: StructType,
```

Comment:
What's the downside of just storing `schemaJson: String` as before? We can validate the given string in `require` if that was the concern. I may have missed other concerns.

Reply:
This PR is a follow-up of #443 and this part belongs to that. Please ignore other commits and see the commit 7e0a6693c0618091d56b0485ae05ed8b9494f31c for the changes of this PR.

I think the downsides are as follows:

1. You have to convert the string to `StructType` every time you need it. You can make a lazy val and `@JsonIgnore` it but then it consumes more memory. These overheads might be slim, but why pay more when you don't have to?
2. Escaped JSON strings are ugly and less readable than normal JSON strings. Could be a minor nuisance when debugging.

Reply:
> 1. You have to convert the string to `StructType` every time you need it. You can make a lazy val and `@JsonIgnore` it but then it consumes more memory. These overheads might be slim, but why pay more when you don't have to?

How much memory are we talking here? :) I guess we are paying for the unnecessary macros-related code? In #443, it seems we can just use `schemaString` and not worry about separating 2.4 vs. 3.0 in the first place, no?



> 2\. Escaped JSON strings are ugly and less readable than normal JSON strings. Could be a minor nuisance when debugging.

Seems minor to me. There are many unescaping tools available online if readability is really an issue while debugging.

Reply:
I will leave it up to you and @sezruby on the final decision. Just checking to see if shim was necessary for this field.

Reply:
Seems code redundancy & complexity for StructTypeConverter outweigh the gain. I think we could introduce these macro when it's really required for expensive conversion/shim (e.g. data file paths?) later.

Reply:
Why use tools when you don't have to? :)

These macros are general utilities. If it has any cost, it is amortized as we use it more.

I'm fine delivering this when more needs arise. The current usage is not strictly necessary; it's just a demonstration.

Reply:
> Why use tools when you don't have to? :)

Because it is brining in unnecessary complexity to the code base.

Reply:
First, it might be "unnecessary" for now but might be necessary in the future. I agree that we can wait until a more strong need arises. Whether something is necessary or unnecessary is subjective and I think it can be necessary.

Second, about complexity - having one more subproject can feel complex for developers unfamiliar with multi-projects SBT builds, but once it's set up, you don't need to know it's there. The macro code may be complex, but the complexity is contained - you don't have to understand the macro code in order to understand the codebase. You can ignore the macro subproject entirely and don't have to use macros for your daily Hyperspace coding.